File: 00019. Remove Nth Node From End of List/00019.c
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/

struct ListNode * removeNthFromEnd(struct ListNode * head, int n) {
    struct ListNode *fast = head, *slow = NULL, *last = NULL;

    for (int i = 0; i < n; ++i) {
        fast = fast->next;
    }

    for (slow = head, last = NULL; fast != NULL; fast = fast->next, slow = slow->next) {
        last = slow;
    }

    if (slow == head) {
        head = head->next;
    } else {
        last->next = slow->next;
    }

    free(slow);
    return head;
}

File: 00019. Remove Nth Node From End of List/00019.go
// https://leetcode.com/problems/remove-nth-node-from-end-of-list

func removeNthFromEnd(head *ListNode, n int) *ListNode {
    var prev *ListNode = nil
    var fast *ListNode = head

    for i := 0; i < n; i++ {
        fast = fast.Next
    }

    if fast == nil {
        return head.Next
    }

    // 1 step for init prev to head
    prev, fast = head, fast.Next

    for fast != nil {
        prev = prev.Next
        fast = fast.Next
    }

    prev.Next = prev.Next.Next
    return head
}
